### PR TITLE
feature: provide a way to backup and restore server and plugin settings in the admin ui

### PR DESCRIPTION
**Pull request by sbender9**
TODO:

- [x] run `npm install` to get plugins installed
- [x] Show some kind of progress in the ui
- [x] add restart button or confirmation 
- [x] make so if you go away and come back to the settings page the progress shows

image

image

image

### CODE REVIEW

File: packages/server-admin-ui/src/views/ServerConfig/Settings.js [hunk 1]
```diff
@@ -32,20 +32,81 @@ class Settings extends Component {
...
+        credentials: 'include',
+        method: 'POST',
+        headers: {
+          'Accept': 'application/json'
+        },
+        body: data
+      })
+        .then(response => {
+          if (response.ok) {
+            alert('The backup was restored. Please restart the server.')
```
Comment (tkurki) [inline packages/server-admin-ui/src/views/ServerConfig/Settings.js]: this should trigger a restart, not ask the user to restart

File: src/serverroutes.js [hunk 2]
```diff
@@ -629,4 +633,65 @@ module.exports = function(app, saveSecurityConfig, getSecurityConfig) {
...
+
+  app.post('/restore', (req, res, next) => {
+    const busboy = new Busboy({ headers: req.headers })
+    busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
+      try {
+        if (!filename.endsWith('.backup')) {
+          res.status(400).send('Invalid backup file')
+          return
+        }
+        const unzipStream = unzipper.Extract({ path: app.config.configPath })
```
Comment (tkurki) [inline src/serverroutes.js]: I think we should sanitize the input somehow, I feel quezy about just unzipping something.
- upper limit to file size
- only known files in the settings base directory?
- i guess plugins settings and data directories can not be effectively sanitized, there can be dragons there

Should we report to the user something user friendly about what the backup contained? 

Should there be a way to see into the backup file without restoring, like a preview of what there is?

File: src/serverroutes.js [hunk 3]
```diff
@@ -629,4 +633,65 @@ module.exports = function(app, saveSecurityConfig, getSecurityConfig) {
       })
     )
   })
+
+  app.post('/restore', (req, res, next) => {
+    const busboy = new Busboy({ headers: req.headers })
+    busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
+      try {
+        if (!filename.endsWith('.backup')) {
+          res.status(400).send('Invalid backup file')
```
Comment (tkurki) [inline src/serverroutes.js]: should the prefix `signalk-` be also checked? A more friendly error message about what's wrong with the file?

File: packages/server-admin-ui/src/views/ServerConfig/Settings.js (hunk 1 above)
Comment (sbender9) [reply packages/server-admin-ui/src/views/ServerConfig/Settings.js]: I started to do that and then changed my mind, I forget why...

File: src/serverroutes.js (hunk 2 above)
Comment (sbender9) [reply src/serverroutes.js]: All sounds good

File: src/serverroutes.js (hunk 3 above)
Comment (sbender9) [reply src/serverroutes.js]: ok

**Pull request merged**